### PR TITLE
Update getUserMedia API

### PR DESCRIPTION
**Pull request by hjheath**
https://developer.mozilla.org/en/docs/Web/API/Navigator/getUserMedia

This fixes a camera bug on firefox and allows us to cope with error situations where `getUserMedia` denies access to the camera (e.g. user does not give permission)

### CODE REVIEW

File: src/react-webcam.js [hunk 1]
```diff
@@ -2,8 +2,7 @@ import React, { Component, PropTypes } from 'react';
 import { findDOMNode } from 'react-dom';
 
 function hasGetUserMedia() {
-  return !!(navigator.getUserMedia || navigator.webkitGetUserMedia ||
-            navigator.mozGetUserMedia || navigator.msGetUserMedia);
+  return !!(navigator.mediaDevices && navigator.mediaDevices.getUserMedia)
```
Comment (rfreitas) [inline src/react-webcam.js]: why did you remove the prefixed checking?

Comment (hjheath) [inline src/react-webcam.js]: The prefixes no longer apply with the new version of `getUserMedia` https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices

Comment (rfreitas) [inline src/react-webcam.js]: but older version browser still need the prefix though

File: src/react-webcam.js [hunk 2]
```diff
@@ -1,9 +1,34 @@
...
+  if (navigator.mediaDevices.getUserMedia === undefined) {
+    navigator.mediaDevices.getUserMedia = function(constraints) {
+
+      // First get hold of the legacy getUserMedia, if present
+      const getUserMedia = navigator.webkitGetUserMedia || navigator.mozGetUserMedia
+
+      // Some browsers just don't implement it - return a rejected promise with an error
+      // to keep a consistent interface
+      if (!getUserMedia) {
+        return Promise.reject(new Error('getUserMedia is not implemented in this browser'))
```
Comment (hjheath) [inline src/react-webcam.js]: TODO: IE does not support Promises, check that this will work

Comment (stephencookdev) [inline src/react-webcam.js]: It doesn't look like you have any polyfilling, so I'd be surprised!

But do you need it? Looks to me like you never use the return statement

Comment (hjheath) [inline src/react-webcam.js]: I think we would hit this in the case of getUserMedia not being implemented at all (i.e. IE 11).

I put some logic in the parent component to detect getUserMedia so we don't even try and use this component if it is not supported.

File: src/react-webcam.js [hunk 3]
```diff
@@ -1,9 +1,34 @@
...
+    navigator.mediaDevices = {}
+  }
+
+  // Some browsers partially implement mediaDevices. We can't just assign an object
+  // with getUserMedia as it would overwrite existing properties.
+  // Here, we will just add the getUserMedia property if it's missing.
+  if (navigator.mediaDevices.getUserMedia === undefined) {
+    navigator.mediaDevices.getUserMedia = function(constraints) {
+
+      // First get hold of the legacy getUserMedia, if present
```
Comment (rfreitas) [inline src/react-webcam.js]: weird indentation here (seems like you removed an if clause

File: src/react-webcam.js (hunk 2 above)
Comment (rfreitas) [inline src/react-webcam.js]: yeah I'm having second thoughts about this polyfill. We should actually have a ponyfill instead, especially considering we are an sdk in an app, changing global is not nice.

File: src/react-webcam.js [hunk 4]
```diff
@@ -1,9 +1,34 @@
...
+  }
+  else if (navigatorGetUserMedia) {
+    getUserMedia = (constraints) => {
+      return new Promise(function(resolve, reject) {
+        navigatorGetUserMedia.call(navigator, constraints, resolve, reject)
+      })
+    }
+  }
+  else {
+    getUserMedia = () =>
```
Comment (rfreitas) [inline src/react-webcam.js]: I would still return `null` here. If promises weren't a compatibility issue this code would be the best due to consistency. But since they are, returning `null` here is a very reasonable tradeoff considering what he get out.

Comment (rfreitas) [inline src/react-webcam.js]: I just checked an IE11 does not have `getUserMedia` and it does not have promises, so it will crash here, won't it?

Comment (rfreitas) [inline src/react-webcam.js]: I'm guessing the parent is checking for `hasGetUserMedia`, but this component can also be defensive and never call this function in case that boolean is `false`

File: src/react-webcam.js [hunk 5]
```diff
@@ -1,9 +1,33 @@
...
+// Adapted from https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia
+// Use a pony fill to avoid editing global objects
+function getUserMediaPonyfill() {
+  let getUserMedia
+  if (mediaDevices) {
+    getUserMedia = navigator.mediaDevices.getUserMedia.bind(navigator.mediaDevices)
+  }
+  else if (navigatorGetUserMedia) {
+    getUserMedia = (constraints) => {
+      return new Promise(function(resolve, reject) {
```
Comment (rfreitas) [inline src/react-webcam.js]: just as style pointer, in a more functional approach one would drop the `let` and just return on every clause.
